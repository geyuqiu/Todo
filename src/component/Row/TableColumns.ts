export const tableColumns = [
	{
		Header: '#',
		accessor: 'id'
	},
	{
		Header: 'Title',
		accessor: 'title',
		className: "text-right"
	},
	{
		Header: 'Completed',
		accessor: 'completed',
	}
];
