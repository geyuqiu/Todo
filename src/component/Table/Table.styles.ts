export const tableStyle = `
	table {
		margin: 0;
		padding: 0;
		width: 100%;
		border-color: gray;
		border-bottom-width: 1px;

		thead {
			th.hasBorder:not(:last-child):after {
				content: "";
		    height: 50%;
		    position: absolute;
		    right: 0;
		    top: 25%;
			}
		}
	}
`;
