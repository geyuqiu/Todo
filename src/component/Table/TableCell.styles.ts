import tw from "twin.macro";

const baseStyle = tw`flex items-center p-5 my-1 transition-colors duration-100`;

const getHeight = () => tw`sm:h-6 h-24`;

const getVariant = (variant: "start" | "middle" | "end"): any => {
	if (variant === "start") {
		return tw`pl-4 -ml-4 rounded-l-lg`;
	}

	if (variant === "end") {
		return tw`pr-4 -mr-4 rounded-r-lg`;
	}
};

export const getStyles = ({ variant }: any) => [baseStyle, getHeight(), getVariant(variant)];
