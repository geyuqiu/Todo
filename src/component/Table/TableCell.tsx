import React from "react";
import {styled} from "twin.macro";
import cn from 'classnames';

import {getStyles} from "./TableCell.styles";

type TableCellProps = {
	variant?: "start" | "middle" | "end";
	className?: string;
	innerClassName?: string;
	isCompact?: boolean;
	children: React.ReactNode;
} & React.HTMLProps<any>;

const TableCellInnerWrapper = styled.div<TableCellProps>(getStyles);

export const TableCell = ({variant, className, innerClassName, isCompact, children, ...props}: TableCellProps) => (
	<td className={cn('border-gray-lightest border', className)} {...props}>
		<TableCellInnerWrapper variant={variant} className={innerClassName} isCompact={isCompact}>
			{children}
		</TableCellInnerWrapper>
	</td>
);

TableCell.defaultProps = {
	variant: "middle",
};
