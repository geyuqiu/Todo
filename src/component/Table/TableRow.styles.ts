import tw from "twin.macro";

import {TableRowFunction} from "./TableRow";

const baseStyle = tw`transition-colors duration-100`;

const getCursorStyles = (onClick?: TableRowFunction): any => (onClick ? tw`cursor-pointer` : "");

const getBorderStyles = (border?: boolean): any =>
	border ? tw`border border-solid border-gray-lightest` : "";

export const getStyles = ({ onClick, border }: any) => {
	const styles = [baseStyle, getBorderStyles(border), getCursorStyles(onClick)];

	return styles;
};
