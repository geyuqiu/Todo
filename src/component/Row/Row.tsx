import React from "react";

import {TableRow} from '../Table/TableRow';
import {TableCell} from '../Table/TableCell';
import {Todo} from './model';

type RowProps = {
	todo: Todo
};

export const Row = ({todo: todo}: RowProps) => {
	return (
		<TableRow className="text-base" border>

			<TableCell>
				{todo.id}
			</TableCell>

			<TableCell>
				{todo.title}
			</TableCell>

			<TableCell>
				{String(todo.completed)}
			</TableCell>

		</TableRow>
	);
}
