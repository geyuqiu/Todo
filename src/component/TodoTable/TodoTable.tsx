import React, {useState} from "react";

import {tableColumns} from '../Row/TableColumns';
import {Row} from '../Row/Row';
import {Table} from '../Table/Table';
import {Todo} from '../Row/model';
import {Filters} from '../Filters/Filters';

type TableProps = {
	todos: Todo[];
	isLoading: boolean;
}

export const TodoTable = ({todos, isLoading}: TableProps) => {
	const [title, setTitle] = useState<string>("");
	const [completed, setCompleted] = useState<string>("");
	return (
		<div className="ml-6 mr-3 sm:mx-8 pt-12">
			<h1 className="font-black mb-4 text-3xl">Todos</h1>
			<Filters title={title} completed={completed} setTitle={setTitle} setCompleted={setCompleted}/>
			{(isLoading) && <p data-testid="Table__Loading">Loading ...</p>}
			{!isLoading && !todos?.length &&
			<NoResults/>
			}
			{!isLoading && todos?.length > 0 &&
			<Table columns={tableColumns} data={todos}
						 titleOutside={title} completedOutside={completed}
			>
				{(todo: Todo) => (
					<Row todo={todo}/>
				)}
			</Table>
			}
		</div>
	);
}

export const NoResults = () =>
	<p data-testid="Table__No_Results" className="bg-error p-4 text-white">No result for the given
	search parameters!</p>
