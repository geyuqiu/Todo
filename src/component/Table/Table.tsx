import React, {useEffect, useMemo} from "react";
import {TableOptions, useFilters, useTable} from "react-table";
import {styled} from "twin.macro";

import {tableStyle} from "./Table.styles";
import {NoResults} from '../TodoTable/TodoTable';

type TableProps = {
	children?: any;
	className?: string;
	data: any[];
	columns: any[];
	initialState?: Record<string, any>;
	titleOutside?: string;
	completedOutside?: string;
};

const TableWrapper = styled.div`
	${tableStyle}
`;

export const Table = ({children, data, columns, className, initialState, titleOutside, completedOutside}: TableProps) => {
	const filterTypes = React.useMemo(() => ({
		text: (rows: any, id: any, filterValue: any) => {
			return rows.filter((row: any) => {
				const rowValue = row.values[id]
				return rowValue !== undefined
					? String(rowValue)
						.toLowerCase()
						.startsWith(String(filterValue).toLowerCase())
					: true
			})
		},
	}), [])

	const {getTableProps, getTableBodyProps, headerGroups, rows, prepareRow,
		// @ts-ignore
		setFilter} = useTable(
		{
			data: useMemo(() => data, [data]),
			columns: useMemo(() => columns, [columns]),
			initialState,
			filterTypes
		} as TableOptions<any>, useFilters
	);

	const renderChildNode = (data: any, index: number) => {
		if (typeof children === "function") {
			return children(data, index);
		}
		return <tr/>;
	};

	useEffect(() => {
		if (titleOutside) setFilter("title", titleOutside);
	}, [titleOutside]);

	useEffect(() => {
		if (completedOutside) setFilter("completed", completedOutside);
	}, [completedOutside]);

	return (
		<>
			{!rows.length && <NoResults/>}
			{rows.length ?
				<TableWrapper {...getTableProps({className})} className="w-full max-w-screen-xl">
					<table cellPadding={0} className="table-auto">
						<thead>
						{headerGroups.map((headerGroup: any, index: number) => (
							<tr className="border" key={index} {...headerGroup.getHeaderGroupProps()}>
								{headerGroup.headers.map((column: any, thIndex: number) => (
									<th key={thIndex}
									    className="group border border-gray-lightest bg-gray-dark text-base text-center select-none m-0 p-3 md:px-5 font-semibold"
									    {...column.getHeaderProps()}
									    data-testid={`table__th--${thIndex}`}
									>{column.render("Header")}</th>
								))}
							</tr>
						))}
						</thead>
						<tbody {...getTableBodyProps()}>
						{rows.map((row: any) => {
							prepareRow(row);
							return {...renderChildNode(row.original, row.index), ...row.getRowProps()};
						})}
						</tbody>
					</table>
				</TableWrapper> : <div/>}
		</>
	);
};

Table.defaultProps = {
	data: [],
	columns: []
};
