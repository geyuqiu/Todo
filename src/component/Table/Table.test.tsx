import {render} from "@testing-library/react";
import React from "react";

import {Table} from "./Table";
import {tableColumns} from '../Row/TableColumns';
import {todos} from '../TodoTable/TodoTable.test';

describe("Table", () => {
	it("should render", () => {
		const {asFragment} = render(
			<Table columns={tableColumns} data={todos}>
				{() => (
					<tr>
						<td>1</td>
						<td>2</td>
					</tr>
				)}
			</Table>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it("should render empty rows if template not provided", () => {
		const {container} = render(<Table columns={tableColumns} data={todos}/>);
		expect(container).toMatchSnapshot();
	});
});
