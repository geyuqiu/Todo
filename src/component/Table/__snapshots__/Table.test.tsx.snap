// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render 1`] = `
<div>
  <div
    class="Table__TableWrapper-xscllg-0 gqOIbh w-full max-w-screen-xl"
    role="table"
  >
    <table
      cellpadding="0"
      class="table-auto"
    >
      <thead>
        <tr
          class="border"
          role="row"
        >
          <th
            class="group border border-gray-lightest bg-gray-dark text-base text-center select-none m-0 p-3 md:px-5 font-semibold"
            colspan="1"
            data-testid="table__th--0"
            role="columnheader"
          >
            Header 1
          </th>
          <th
            class="group border border-gray-lightest bg-gray-dark text-base text-center select-none m-0 p-3 md:px-5 font-semibold"
            colspan="1"
            data-testid="table__th--1"
            role="columnheader"
          >
            Header 2
          </th>
          <th
            class="group border border-gray-lightest bg-gray-dark text-base text-center select-none m-0 p-3 md:px-5 font-semibold"
            colspan="1"
            data-testid="table__th--2"
            role="columnheader"
          >
            Header 3
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr>
          <td>
            1
          </td>
          <td>
            2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should render empty rows if template not provided 1`] = `
<div>
  <div
    class="Table__TableWrapper-xscllg-0 gqOIbh w-full max-w-screen-xl"
    role="table"
  >
    <table
      cellpadding="0"
      class="table-auto"
    >
      <thead>
        <tr
          class="border"
          role="row"
        >
          <th
            class="group border border-gray-lightest bg-gray-dark text-base text-center select-none m-0 p-3 md:px-5 font-semibold"
            colspan="1"
            data-testid="table__th--0"
            role="columnheader"
          >
            Header 1
          </th>
          <th
            class="group border border-gray-lightest bg-gray-dark text-base text-center select-none m-0 p-3 md:px-5 font-semibold"
            colspan="1"
            data-testid="table__th--1"
            role="columnheader"
          >
            Header 2
          </th>
          <th
            class="group border border-gray-lightest bg-gray-dark text-base text-center select-none m-0 p-3 md:px-5 font-semibold"
            colspan="1"
            data-testid="table__th--2"
            role="columnheader"
          >
            Header 3
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr />
      </tbody>
    </table>
  </div>
</div>
`;
